src/Evolution/DiscontinuousGalerkin/SlopeLimiters/HwenoModifiedSolution.cpp

```diff
@@ -0,0 +1,340 @@
+// See LICENSE.txt for details.
+
+#include "Evolution/DiscontinuousGalerkin/SlopeLimiters/HwenoModifiedSolution.hpp"
+
+#include <array>
+#include <bitset>
+#include <cstddef>
+#include <ostream>
+#include <vector>
+
+#include "DataStructures/IndexIterator.hpp"
+#include "Domain/Direction.hpp"
+#include "Domain/Element.hpp"  // IWYU pragma: keep
+#include "Domain/Mesh.hpp"     // IWYU pragma: keep
+#include "Domain/Side.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.hpp"
+#include "NumericalAlgorithms/Interpolation/RegularGridInterpolant.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+
+// Compute the VolumeDim-dimensional quadrature weights, by taking the tensor
+// product of the 1D quadrature weights in each logical dimension of the Mesh.
+template <size_t VolumeDim>
+DataVector volume_quadrature_weights(const Mesh<VolumeDim>& mesh) noexcept {
+  std::array<DataVector, VolumeDim> quadrature_weights_1d{};
+  for (size_t d = 0; d < VolumeDim; ++d) {
+    gsl::at(quadrature_weights_1d, d) =
+        Spectral::quadrature_weights(mesh.slice_through(d));
+  }
+
+  DataVector result(mesh.number_of_grid_points(), 1.);
+  for (IndexIterator<VolumeDim> s(mesh.extents()); s; ++s) {
+    result[s.collapsed_index()] = quadrature_weights_1d[0][s()[0]];
+    for (size_t d = 1; d < VolumeDim; ++d) {
+      result[s.collapsed_index()] *= gsl::at(quadrature_weights_1d, d)[s()[d]];
+    }
+  }
+  return result;
+}
+
+// Compute the matrix that interpolates data in VolumeDim dimensions from
+// source_mesh to the grid points of target_mesh, by taking the tensor product
+// of the 1D interpolation matrices in each logical dimension of the meshes.
+//
+// Note: typically it is more efficient to interpolate from source_mesh to
+// target_mesh in one logical direction at a time, and to avoid computing this
+// matrix. However, the HWENO algorithm requires constructing this matrix
+// explicitly.
+//
+// In particular, for this HWENO application:
+// - source_mesh is the mesh of local element.
+// - target_mesh is the mesh of the neighboring element in direction. Note that
+//   there can only be one neighbor in this direction, and this is checked in
+//   the call to `neighbor_grid_points_in_local_logical_coordinates`.
+// - rectilinear elements are assumed.
+template <size_t VolumeDim>
+Matrix volume_interpolation_matrix(
+    const Mesh<VolumeDim>& source_mesh, const Mesh<VolumeDim>& target_mesh,
+    const Element<VolumeDim>& element,
+    const Direction<VolumeDim>& direction) noexcept {
+  // The grid points of source_mesh and target_mesh must be in the same
+  // coordinates to construct the interpolation matrix. Here we get the points
+  // of target_mesh in the local logical coordinates.
+  const auto target_1d_coords =
```

Comment:
Yes, you'd lose the information that these are 1D coords? 😛

This data is an array where each element is a vector {xmin, ..., xmax} (same in y, same in z). The full coordinates are obtained by picking one value from each vector. In other words, supposing a 3D mesh with N grid points per dimension, we are storing 3N pieces of data rather than 3 * N^3.

I'd like to keep this distinction in the variable name. However, I welcome suggestions for clearer names 🙂